iceoryx_hoofs/utility/include/iox/detail/convert.inl

```diff
@@ -58,384 +60,404 @@ convert::toString(const Source& t) noexcept
+{
     if (call.has_error())
     {
-        return false;
+        return iox::nullopt;
     }
 
-    if (call->value > std::numeric_limits<int32_t>::max() || call->value < std::numeric_limits<int32_t>::min())
+    if (!check_edge_case<TargetType>(errno_cache, end_ptr, v, call->value))
     {
-        IOX_LOG(DEBUG, call->value << " is out of range, int32_t overflow");
-        return false;
+        return iox::nullopt;
     }
 
-    dest = static_cast<int32_t>(call->value);
-    return true;
+    // clean errno
+    errno = 0;
+
+    return iox::optional<TargetType>(static_cast<TargetType>(call->value));
 }
 
-template <>
-inline bool convert::fromString<int16_t>(const char* v, int16_t& dest) noexcept
+template <typename SourceType>
+inline bool convert::is_valid_input(const char* end_ptr, const char* v, const SourceType& source_val) noexcept
 {
-    if (!stringIsNumberWithErrorMessage(v, NumberType::INTEGER))
+    // invalid string
+    if (v == end_ptr && source_val == 0)
     {
+        IOX_LOG(DEBUG, "invalid input");
         return false;
     }
 
-    auto call = IOX_POSIX_CALL(strtol)(v, nullptr, STRTOULL_BASE).failureReturnValue(LONG_MAX, LONG_MIN).evaluate();
-    if (call.has_error())
+    // end_ptr is not '\0' which means conversion failure at end_ptr
+    if (end_ptr != nullptr && v != end_ptr && *end_ptr != '\0')
     {
+        IOX_LOG(DEBUG, "conversion failed at " << end_ptr - v << " : " << *end_ptr);
         return false;
     }
 
-    if (call->value > std::numeric_limits<int16_t>::max() || call->value < std::numeric_limits<int16_t>::min())
-    {
-        IOX_LOG(DEBUG, call->value << " is out of range, int16_t overflow");
-        return false;
-    }
-
-    dest = static_cast<int16_t>(call->value);
     return true;
 }
 
-template <>
-inline bool convert::fromString<int8_t>(const char* v, int8_t& dest) noexcept
+template <typename SourceType>
+inline bool convert::is_valid_errno(decltype(errno) errno_cache, const char* v, const SourceType& source_val) noexcept
 {
-    if (!stringIsNumberWithErrorMessage(v, NumberType::INTEGER))
+    // check errno
+    if (errno_cache == ERANGE)
     {
+        if constexpr (std::is_floating_point_v<SourceType>)
+        {
+            /// @todo iox-#2055
+            // This could be a case of 'inf' or 'nan' for floating points,
+            // yet the use of 'failureReturnValue' effectively prevents the occurrence of 'inf'.
+            // In such instances, due to 'result.has_error' being true,
+            // 'from_string' will return prematurely.
+            // Consequently, this function will remain uncalled.
+
+            // currently, we still treat them as conversion error
+            // but they should be treated as potential result (instead of error)?
+            IOX_LOG(DEBUG,
+                    "ERANGE triggered during conversion of string: '" << v << "'. The result is: " << source_val);
+        }
```

Comment:
I think with `ignoreErrnos` instead of `suppressErrorMessagesForErrnos` this can be removed and `is_within_range` can check for `inf` and `nan` and abort for a `signaling nan`. The check for the `signaling nan` should probably be the first before the floating point value is read to prevent a floating point exception.

And we also need a test to ensure that a `signaling nan` is converted to a `nullopt`

Reply:
I've encountered a few issues:

1. It seems that `strtof` and `strtod` never return a signaling NaN? In this case, is it necessary to test for it? If it is, we might need to think of a way to simulate the results produced by `IOX_POSIX_CALL`.
2. Whether a `long double` is considered a signaling NaN is platform-dependent. We need to take into account these platform dependencies. Moreover, the choice of compiler options can also change the length of `long double`, thereby affecting the determination of the signaling bit.

   - [GNU C's x86 flags might affect the length of long double](https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html#:~:text=not%20binary%20compatible.-,%2Dm96bit%2Dlong%2Ddouble,with%20programs%20compiled%20without%20the%20switch%20are%20not%20binary%20compatible.)
   - [The length of long double is platform-dependent](https://en.wikipedia.org/wiki/Long_double#:~:text=With%20the%20GNU,%5B20%5D)

    How should we deal with this complex phenomenon? Currently, my thought is to exclude the NaN case for `long double`(always consider it as a quiet NaN) and only implement a general method for detecting sNaN (`is_signaling_nan`) for `float` and `double`.
3. Different architectures have inconsistent definitions for the is_quiet bit. For example, PA-RISC defines it as is_signaling, not is_quiet, as mentioned in [this link](https://en.wikipedia.org/wiki/NaN#:~:text=compatibility%22.%5B12%5D-,Encoding,-%5Bedit%5D). What would be your advice in such situations?

Reply:
Oh no, I did it again :sweat_smile: 

You are totally right, there is no string representation for a signaling nan and therefore it does not need to be checked and tested. Sorry, I have overshot the mark.

Reply:
Please don't be sorry! Actually, I don't know what the difference is between quiet NaN and signaling NaN. Your suggestion led me to do some research on them, and I think I gained some benefits from it :) I will simplify the happy path for NaN as you just suggested!